File: frontend/src/api/types/doeModelWorkflow.ts
export interface DoeModelSelectionOptions {
  method: "none" | "backward_elimination";
  alpha_to_remove: number;
  hierarchy_policy: "strong";
  saturated_start_policy: "pool_smallest_adjusted_ss";
  display_step_details: boolean;
}

export interface DoeModelSelectionStep {
  step: number;
  phase: "initial_full_model" | "initial_pooling" | "backward_elimination";
  active_term_ids: string[];
  removed_term_id: string | null;
  removal_df: number | null;
  removal_adjusted_ss: number | null;
  removal_p_value: number | null;
  stop_reason: string | null;
  coefficients: Array<{ column_index: number; coefficient: number }>;
  sse: number;
  residual_df: number;
  residual_standard_error: number | null;
  r_squared: number;
  adjusted_r_squared: number | null;
  press: number | null;
  predicted_r_squared: number | null;
}

export interface DoeModelSelectionResult extends DoeModelSelectionOptions {
  tie_break_policy: string;
  initial_model_saturated: boolean;
  initial_parameter_count: number;
  initial_residual_df: number;
  target_pool_count: number;
  initial_term_ids: string[];
  pooled_term_ids: string[];
  removed_term_ids: string[];
  final_term_ids: string[];
  fixed_term_ids: string[];
  stop_reason: string;
  steps: DoeModelSelectionStep[];
  post_selection_inference: "exploratory" | "not_selected";
}

export interface DoeFinalCoefficient {
  term_id: string;
  label: string;
  kind: string;
  factor_names: string[];
  coefficient: number;
  effect: number | null;
  standard_error: number | null;
  t_statistic: number | null;
  p_value: number | null;
  vif: number | null;
  ci_lower: number | null;
  ci_upper: number | null;
  status: "structural" | "retained";
}

export interface DoePredictionFeature {
  term_id: string;
  label: string;
  kind: "intercept" | "main_effect" | "interaction" | "center_curvature" | "block";
  factor_names: string[];
  level_indices: number[];
  block_index: number | null;
}

export interface DoePredictionBasis {
  schema_version: 1;
  coding: "coded" | "treatment";
  features: DoePredictionFeature[];
  coefficients: number[];
  xtx_inverse: number[][];
  residual_mean_square: number | null;
  residual_df: number;
  confidence_level: number;
  block_levels: Array<number | null>;
  center_settings: Array<Record<string, number>>;
}

export interface DoeResidualView {
  n: number;
  histogram: Array<{ lower: number; upper: number; count: number }>;
  qq_points: Array<{ run_order: number; theoretical_quantile: number; residual: number }>;
  reference_line: { slope: number; intercept: number } | null;
  points: Array<{ run_order: number; fitted: number; residual: number }>;
}

export interface DoeFinalModelWorkflow {
  anova_groups?: Array<{ kind: string; label: string; df: number; adjusted_ss: number; mean_square: number | null; f_statistic: number | null; p_value: number | null; terms: Array<{ label: string; df: number; adjusted_ss: number; mean_square: number | null; f_statistic: number | null; p_value: number | null }> }>;
  equation: { scale: "coded" | "treatment"; response_name: string; intercept: number; terms: DoeFinalCoefficient[]; display_equation: string };
  coded_coefficients: DoeFinalCoefficient[];
  prediction_basis: DoePredictionBasis;
  press: number | null;
  predicted_r_squared: number | null;
  residual_plots: {
    raw: DoeResidualView;
    standardized: DoeResidualView;
    points: Array<{ run_order: number; observed: number; fitted: number; residual: number; standardized_residual: number | null; leverage: number; cooks_distance: number | null }>;
    n_total: number; point_limit: number; truncated: boolean;
  };
  factorial_plots: {
    cells: Array<{ settings: Record<string, number>; fitted_mean: number; data_mean: number | null; n: number }>;
    available: boolean; reason: string | null; marginalization: "equal_level_and_block_weights";
  };
  warnings: string[];
}

export interface DoePredictionRowRequest {
  row_id: string;
  factor_settings: Record<string, number | string>;
  block_index?: number | null;
}
export interface DoePredictionPreflightRequest {
  rows: DoePredictionRowRequest[];
  confidence_level: number;
}
export interface DoePredictionPreflightResponse {
  schema_version: 1; design_id: string; analysis_id: string; source_analysis_sha256: string;
  preflight_sha256: string; valid: boolean; row_count: number;
  issues: Array<{ row_id: string; code: string; factor_name: string | null }>;
  warnings: string[];
}
export interface DoePredictionInterval { lower: number; upper: number; level: number }
export interface DoePredictionResponse {
  schema_version: 1; prediction_id: string; design_id: string; analysis_id: string;
  response_revision_id: string; response_revision_sha256: string; source_analysis_sha256: string;
  created_at: string;
  rows: Array<{ row_id: string; factor_settings: Record<string, number | string>; block_index: number | null;
    fitted_mean: number; standard_error_fit: number | null; mean_confidence_interval: DoePredictionInterval | null;
    individual_prediction_interval: DoePredictionInterval | null; interval_unavailability_reason: string | null; domain_status: "in_domain" }>;
  warnings: string[];
}
export interface DoeAnalysisAssetDescriptor {
  asset_id: string; analysis_id: string; kind: "prediction" | "html_report"; schema_version: 1;
  locale: "en" | "ko" | null; source_analysis_sha256: string; sha256: string; media_type: string;
  size_bytes: number; created_at: string;
}
export interface DoeAnalysisAssetList { analysis_id: string; items: DoeAnalysisAssetDescriptor[] }
export interface DoeAnalysisDeletionPreflight {
  analysis_id: string; prediction_count: number; report_count: number; deletion_manifest_sha256: string;
}

File: backend/app/api/v1/schemas/doe_model_workflow.py
from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, FiniteFloat, field_validator


class WorkflowModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class DoeModelSelectionOptions(WorkflowModel):
    method: Literal["none", "backward_elimination"] = "none"
    alpha_to_remove: FiniteFloat = Field(default=0.05, gt=0, lt=1)
    hierarchy_policy: Literal["strong"] = "strong"
    saturated_start_policy: Literal["pool_smallest_adjusted_ss"] = "pool_smallest_adjusted_ss"
    display_step_details: bool = True


class SelectionStepCoefficient(WorkflowModel):
    column_index: int
    coefficient: FiniteFloat


class DoeModelSelectionStep(WorkflowModel):
    step: int
    phase: Literal["initial_full_model", "initial_pooling", "backward_elimination"]
    active_term_ids: list[str]
    removed_term_id: str | None
    removal_df: int | None
    removal_adjusted_ss: FiniteFloat | None
    removal_p_value: FiniteFloat | None
    stop_reason: str | None
    coefficients: list[SelectionStepCoefficient]
    sse: FiniteFloat
    residual_df: int
    residual_standard_error: FiniteFloat | None
    r_squared: FiniteFloat
    adjusted_r_squared: FiniteFloat | None
    press: FiniteFloat | None
    predicted_r_squared: FiniteFloat | None


class DoeModelSelectionResult(DoeModelSelectionOptions):
    tie_break_policy: str
    initial_model_saturated: bool
    initial_parameter_count: int
    initial_residual_df: int
    target_pool_count: int
    initial_term_ids: list[str]
    pooled_term_ids: list[str]
    removed_term_ids: list[str]
    final_term_ids: list[str]
    fixed_term_ids: list[str]
    stop_reason: str
    steps: list[DoeModelSelectionStep]
    post_selection_inference: Literal["exploratory", "not_selected"]


class DoeFinalCoefficient(WorkflowModel):
    term_id: str
    label: str
    kind: str
    factor_names: list[str]
    coefficient: FiniteFloat
    effect: FiniteFloat | None
    standard_error: FiniteFloat | None
    t_statistic: FiniteFloat | None
    p_value: FiniteFloat | None
    vif: FiniteFloat | None
    ci_lower: FiniteFloat | None
    ci_upper: FiniteFloat | None
    status: Literal["structural", "retained"]


class DoeFinalEquation(WorkflowModel):
    scale: Literal["coded", "treatment"]
    response_name: str
    intercept: FiniteFloat
    terms: list[DoeFinalCoefficient]
    display_equation: str


class DoePredictionFeature(WorkflowModel):
    term_id: str
    label: str
    kind: Literal["intercept", "main_effect", "interaction", "center_curvature", "block"]
    factor_names: list[str]
    level_indices: list[int]
    block_index: int | None


class DoePredictionBasis(WorkflowModel):
    schema_version: Literal[1]
    coding: Literal["coded", "treatment"]
    features: list[DoePredictionFeature]
    coefficients: list[FiniteFloat]
    xtx_inverse: list[list[FiniteFloat]]
    residual_mean_square: FiniteFloat | None
    residual_df: int
    confidence_level: FiniteFloat
    block_levels: list[int | None]
    center_settings: list[dict[str, FiniteFloat]]


class DoeResidualHistogramBin(WorkflowModel):
    lower: FiniteFloat
    upper: FiniteFloat
    count: int


class DoeResidualQqPoint(WorkflowModel):
    run_order: int
    theoretical_quantile: FiniteFloat
    residual: FiniteFloat


class DoeResidualReferenceLine(WorkflowModel):
    slope: FiniteFloat
    intercept: FiniteFloat


class DoeResidualScatterPoint(WorkflowModel):
    run_order: int
    fitted: FiniteFloat
    residual: FiniteFloat


class DoeResidualView(WorkflowModel):
    n: int
    histogram: list[DoeResidualHistogramBin]
    qq_points: list[DoeResidualQqPoint]
    reference_line: DoeResidualReferenceLine | None
    points: list[DoeResidualScatterPoint]


class DoeFinalDiagnosticPoint(DoeResidualScatterPoint):
    observed: FiniteFloat
    standardized_residual: FiniteFloat | None
    leverage: FiniteFloat
    cooks_distance: FiniteFloat | None


class DoeFinalResidualPlots(WorkflowModel):
    raw: DoeResidualView
    standardized: DoeResidualView
    points: list[DoeFinalDiagnosticPoint]
    n_total: int
    point_limit: int
    truncated: bool


class DoeFactorialPlotCell(WorkflowModel):
    settings: dict[str, FiniteFloat]
    fitted_mean: FiniteFloat
    data_mean: FiniteFloat | None
    n: int


class DoeFinalFactorialPlots(WorkflowModel):
    cells: list[DoeFactorialPlotCell]
    available: bool
    reason: str | None
    marginalization: Literal["equal_level_and_block_weights"]


class DoeFinalAnovaRow(WorkflowModel):
    label: str
    df: int
    adjusted_ss: FiniteFloat
    mean_square: FiniteFloat | None
    f_statistic: FiniteFloat | None
    p_value: FiniteFloat | None


class DoeFinalAnovaGroup(DoeFinalAnovaRow):
    kind: str
    terms: list[DoeFinalAnovaRow]


class DoeFinalModelWorkflow(WorkflowModel):
    equation: DoeFinalEquation
    coded_coefficients: list[DoeFinalCoefficient]
    anova_groups: list[DoeFinalAnovaGroup] = Field(default_factory=list)
    prediction_basis: DoePredictionBasis
    press: FiniteFloat | None
    predicted_r_squared: FiniteFloat | None
    residual_plots: DoeFinalResidualPlots
    factorial_plots: DoeFinalFactorialPlots
    warnings: list[str]


class DoePredictionRowRequest(WorkflowModel):
    row_id: str = Field(min_length=1, max_length=80)
    factor_settings: dict[str, FiniteFloat | str]
    block_index: int | None = None

    @field_validator("factor_settings", mode="before")
    @classmethod
    def bounded_settings(cls, value: object) -> object:
        if not isinstance(value, dict) or not 1 <= len(value) <= 10:
            raise ValueError("doe_factorial_prediction_factor_set_invalid")
        for name, setting in value.items():
            if not isinstance(name, str) or len(name) > 120:
                raise ValueError("doe_factorial_prediction_factor_set_invalid")
            if isinstance(setting, bool) or not isinstance(setting, str | int | float):
                raise ValueError("doe_factorial_prediction_input_invalid")
            if isinstance(setting, str) and (
                len(setting) > 120 or any(ord(char) < 32 for char in setting)
            ):
                raise ValueError("doe_factorial_prediction_input_invalid")
        return value


class DoePredictionPreflightRequest(WorkflowModel):
    rows: list[DoePredictionRowRequest] = Field(min_length=1, max_length=256)
    confidence_level: FiniteFloat = Field(default=0.95, gt=0, lt=1)


class DoePredictionCreateRequest(DoePredictionPreflightRequest):
    expected_preflight_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")


class DoePredictionRowIssue(WorkflowModel):
    row_id: str
    code: str
    factor_name: str | None = None


class DoePredictionPreflightResponse(WorkflowModel):
    schema_version: Literal[1] = 1
    design_id: UUID
    analysis_id: UUID
    source_analysis_sha256: str
    preflight_sha256: str
    valid: bool
    row_count: int
    issues: list[DoePredictionRowIssue]
    warnings: list[str]


class DoePredictionInterval(WorkflowModel):
    lower: FiniteFloat
    upper: FiniteFloat
    level: FiniteFloat


class DoePredictedRow(WorkflowModel):
    row_id: str
    factor_settings: dict[str, FiniteFloat | str]
    block_index: int | None
    fitted_mean: FiniteFloat
    standard_error_fit: FiniteFloat | None
    mean_confidence_interval: DoePredictionInterval | None
    individual_prediction_interval: DoePredictionInterval | None
    interval_unavailability_reason: str | None
    domain_status: Literal["in_domain"] = "in_domain"


class DoePredictionResponse(WorkflowModel):
    schema_version: Literal[1] = 1
    prediction_id: UUID
    design_id: UUID
    analysis_id: UUID
    response_revision_id: UUID
    response_revision_sha256: str
    source_analysis_sha256: str
    created_at: str
    rows: list[DoePredictedRow]
    warnings: list[str]


class DoeAnalysisAssetDescriptor(WorkflowModel):
    asset_id: UUID
    analysis_id: UUID
    kind: Literal["prediction", "html_report"]
    schema_version: Literal[1]
    locale: Literal["en", "ko"] | None
    source_analysis_sha256: str
    sha256: str
    media_type: str
    size_bytes: int
    created_at: str


class DoeAnalysisAssetList(WorkflowModel):
    analysis_id: UUID
    items: list[DoeAnalysisAssetDescriptor]


class DoeAnalysisAssetDeleteRequest(WorkflowModel):
    confirmation_asset_id: UUID
    expected_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")


class DoeAnalysisAssetDeleteResponse(WorkflowModel):
    asset_id: UUID
    deleted: Literal[True] = True


class DoeAnalysisHtmlExportRequest(WorkflowModel):
    locale: Literal["en", "ko"] = "ko"


class DoeAnalysisDeletionPreflight(WorkflowModel):
    analysis_id: UUID
    prediction_count: int
    report_count: int
    deletion_manifest_sha256: str


class DoeAnalysisDeleteRequest(WorkflowModel):
    confirmation_analysis_id: UUID
    expected_deletion_manifest_sha256: str = Field(pattern=r"^[0-9a-f]{64}$")
